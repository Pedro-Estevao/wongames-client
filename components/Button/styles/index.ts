'use client';

import styled, { css } from 'styled-components';
import { ButtonProps } from '..';

type ContainerModifiersProps = {
	hasIcon: boolean;
} & Pick<ButtonProps, 'size' | 'fullWidth'>;

const containerModifiers = {
	withIcon: () => css`
		display: flex;
		align-items: center;
		justify-content: center;

		svg {
			width: 1.5rem;

			& + span {
				margin-left: 0.8rem;
			}
		}
	`
};

export const ButtonContainer = styled.button<ContainerModifiersProps>`
	${({ hasIcon }) => css`
		${!!hasIcon && containerModifiers.withIcon};
	`}
`;
