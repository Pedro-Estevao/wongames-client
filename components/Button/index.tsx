'use client';

import React, { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';
import * as S from './styles';

type ButtonTypes =
	| AnchorHTMLAttributes<HTMLAnchorElement>
	| ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
	size?: 'small' | 'medium' | 'large';
	fullWidth?: boolean;
	icon?: React.ReactNode;
	as?: React.ElementType;
} & ButtonTypes;

const Button = ({
	children,
	icon,
	size = 'medium',
	fullWidth = false,
	...props
}: ButtonProps) => {
	return (
		<S.ButtonContainer
			className={`inline-flex items-center justify-center bg-gradient-to-b from-orange to-pink text-[white] border-0 rounded-4 p-xxsmall no-underline
                ${
									size === 'small'
										? 'h-[3rem] text-xsmall'
										: size === 'medium'
										? 'h-[4rem] text-small px-medium py-xxsmall'
										: 'h-[5rem] text-medium px-xlarge py-xxsmall'
								}
                ${fullWidth ? 'w-full' : 'w-fit'}
            `}
			hasIcon={!!icon}
			{...props}
		>
			{!!icon && icon}
			{!!children && <span className="leading-normal">{children}</span>}
		</S.ButtonContainer>
	);
};

export default Button;
