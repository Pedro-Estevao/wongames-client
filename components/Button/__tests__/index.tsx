import { render } from '@testing-library/react';
import Button from '@/components/Button';
// import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
// import { faCartShopping } from "@fortawesome/free-solid-svg-icons";

describe('<Button />', () => {
	it('should render the heading', () => {
		const { container } = render(<Button />);

		// expect(screen.getByRole('button', { name: /Buy now/i })).toBeInTheDocument();

		expect(container.firstChild).toMatchSnapshot();
	});

	// it('should render the medium size by default', () => {
	//     const { container } = render(<Button>Buy now</Button>)

	//     expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
	//         height: '4rem',
	//         padding: '0.8rem 3.2rem',
	//         'font-size': '1.4rem'
	//     });

	//     expect(container.firstChild).toMatchSnapshot();
	// });

	// it('should render the small size', () => {
	//     render(<Button size="small">Buy now</Button>);

	//     expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
	//         'font-size': '1.2rem',
	//         height: '3rem',
	//     });
	// });

	// it('should render the large size', () => {
	//     render(<Button size="large">Buy now</Button>);

	//     expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
	//         'font-size': '1.6rem',
	//         height: '6rem',
	//         padding: '0.8rem 4.8rem',
	//     });
	// });

	// it('should render a fullWidth version', () => {
	//     render(<Button fullWidth>Buy now</Button>);

	//     expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
	//         width: '100%'
	//     });
	// });

	// it('should render an icon version', () => {
	//     render(<Button icon={<FontAwesomeIcon icon={faCartShopping} data-testid="icon" />}>Buy now</Button>);

	//     expect(screen.getByText(/buy now/i)).toBeInTheDocument();
	//     expect(screen.getByTestId('icon')).toBeInTheDocument();
	// });
});
