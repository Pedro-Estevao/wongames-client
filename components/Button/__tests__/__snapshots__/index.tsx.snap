// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the heading 1`] = `
<button
  class="bg-gradient-to-b from-orange to-pink text-[white] border-0 rounded-4 p-xxsmall
                h-[4rem] text-small px-medium py-xxsmall
                w-fit
            "
/>
`;
